Fix #10511: Delay 'go to nearest depot' orders

SamuXarick opened the pull request:
## Motivation / Problem
https://github.com/OpenTTD/OpenTTD/issues/10511 

A lone Road Vehicle with a "go to nearest depot" order on a large road network and no depot nearby causes constant calls to the pathfinder every tick, hindering game performance.

This is happening in `ProcessOrders` and not in `IndividualRoadVehicleController` as it's been in the previous reported cases with similar symptoms. I think that the ideal solution here would be to call the pathfinder upon reaching the end of a segment, as that would in theory limit the number of calls significantly. But in this case, it really wants to know what's the destination of the order before handling it over to the `IndividualRoadVehicleController`.

I don't know how to do that, so I thought of a different approach.
<!--
Describe here shortly
* For bug fixes:
    * What problem does this solve?
    * If there is already an issue, link the issue, otherwise describe the problem here.
* For features or gameplay changes:
    * What was the motivation to develop this feature?
    * Does this address any problem with the gameplay or interface?
    * Which group of players do you think would enjoy this feature?
-->


## Description
Delay the nearest depot order search for a day if the destination has been set to zero, which happens when it has already attempted to do so and failed to find a valid destination.

Closes #10511.
<!--
Describe here shortly
* For bug fixes:
    * How is the problem solved?
* For features or gameplay changes:
    * What does this feature do?
    * How does it improve/solve the situation described under 'motivation'.
-->


## Limitations
It may break other "legitimate" uses for `v->dest_tile == 0`

<!--
Describe here
* Is the problem solved in all scenarios?
* Is this feature complete? Are there things that could be added in the future?
* Are there things that are intentionally left out?
* Do you know of a bug or corner case that does not work?
-->


## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR touches english.txt or translations? Check the [guidelines](https://github.com/OpenTTD/OpenTTD/blob/master/docs/eints.md)
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, game_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

glx22 reviewed (dismissed):
Squash required, but final diff looks good to me.

SamuXarick committed:
Fix #10511: Delay 'go to nearest depot' orders

Delay the nearest depot order search for a day if the vehicle can't find its destination, which happens when it has already attempted to do so and failed to find a valid destination.

2TallTyler commented on src/order_cmd.cpp:
```diff
@@ -1961,34 +1961,37 @@ bool UpdateOrderDest(Vehicle *v, const Order *order, int conditional_depth, bool
-							/* The aircraft is now heading for a different hangar than the next in the orders */
-							AircraftNextAirportPos_and_Order(a);
+				/* Try to delay the search for a bit if the previous destination is invalid. */
+				if (v->dest_tile != 0 || TimerGameTick::counter % Ticks::DAY_TICKS == v->index % Ticks::DAY_TICKS) {
```
~~You could save some math with `TimerGameCalendar::date_fract == v->index`~~

glx22 replied:
Would still need a modulo for `v->index` I think.

2TallTyler replied:
Oh, right. `TimerGameCalendar::date_fract == (v->index % Ticks::DAY_TICKS)` ?

PeterN commented on src/order_cmd.cpp:
```diff
@@ -1961,34 +1961,37 @@ bool UpdateOrderDest(Vehicle *v, const Order *order, int conditional_depth, bool
-							/* The aircraft is now heading for a different hangar than the next in the orders */
-							AircraftNextAirportPos_and_Order(a);
+				/* Try to delay the search for a bit if the previous destination is invalid. */
+				if (v->dest_tile != 0 || TimerGameCalendar::date_fract == (v->index % Ticks::DAY_TICKS)) {
```
Please explain, as a comment, why `v->index` is involved in the timing decision.

SamuXarick replied:
I was looking for a way to spread the calls to pathfinder evenly and not concentrate all of them in the same tick. `TimerGameCalendar::date_fract == (v->index % Ticks::DAY_TICKS)` seemed to be a good approach. As each vehicle has a unique index, there will only be a repeat every 74 vehicles in the pool reducing the chances of that happening.

2TallTyler commented on src/order_cmd.cpp:
```diff
@@ -1961,34 +1961,37 @@ bool UpdateOrderDest(Vehicle *v, const Order *order, int conditional_depth, bool
-						if (a->state == FLYING && a->targetairport != closestDepot.destination) {
-							/* The aircraft is now heading for a different hangar than the next in the orders */
-							AircraftNextAirportPos_and_Order(a);
+				/* Try to delay the search for a bit if the previous destination is invalid. */
```
A possible answer to @PeterN's request. 🙂 

```suggestion
				/* If the vehicle can't find its destination, delay its next search.
				 * In case many vehicles are in this state, use the vehicle index to spread out pathfinder calls.*/
```

glx22 commented on src/order_cmd.cpp:
```diff
@@ -1961,34 +1961,38 @@ bool UpdateOrderDest(Vehicle *v, const Order *order, int conditional_depth, bool
-							AircraftNextAirportPos_and_Order(a);
+				/* If the vehicle can't find its destination, delay its next search.
+				 * In case many vehicles are in this state, use the vehicle index to spread out pathfinder calls. */
+				if (v->dest_tile != 0 || TimerGameCalendar::date_fract == (v->index % Ticks::DAY_TICKS)) {
```
```suggestion
				if (v->dest_tile == 0 && TimerGameCalendar::date_fract != (v->index % Ticks::DAY_TICKS)) break;
```
should work too without needing to touch indentation.
(Maybe `return false` instead of break, but I didn't analyse all the function)
